components/script/dom/gamepad.rs

```diff
@@ -0,0 +1,189 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use core::nonzero::NonZero;
+use core::ops::Deref;
+use dom::bindings::cell::DOMRefCell;
+use dom::bindings::codegen::Bindings::GamepadBinding;
+use dom::bindings::codegen::Bindings::GamepadBinding::GamepadMethods;
+use dom::bindings::inheritance::Castable;
+use dom::bindings::js::{JS, MutNullableJS, Root};
+use dom::bindings::num::Finite;
+use dom::bindings::reflector::{DomObject, Reflector, reflect_dom_object};
+use dom::bindings::str::DOMString;
+use dom::event::Event;
+use dom::eventtarget::EventTarget;
+use dom::gamepadbuttonlist::GamepadButtonList;
+use dom::gamepadevent::{GamepadEvent, GamepadEventType};
+use dom::globalscope::GlobalScope;
+use dom::vrpose::VRPose;
+use dom_struct::dom_struct;
+use js::jsapi::{Heap, JSContext, JSObject};
+use js::typedarray::{Float64Array, CreateWith};
+use std::cell::Cell;
+use std::ptr;
+use webvr_traits::{WebVRGamepadData, WebVRGamepadHand, WebVRGamepadState};
+
+#[dom_struct]
+pub struct Gamepad {
+    reflector_: Reflector,
+    gamepad_id: Cell<u64>,
+    id: DOMRefCell<String>,
+    index: Cell<u32>,
```

Comment:
Now it's mutable to ensure that the index is correct if more Gamepad providers are added in the future.